lib/server-builder.js

```diff
@@ -0,0 +1,161 @@
+import { SubProcess } from 'teen_process';
+import { fs, logger, system } from 'appium-support';
+import log from './logger';
+import path from 'path';
+import { EOL } from 'os';
+
+const IGNORED_ERRORS = [];
+const GRADLE_VERSION_KEY = 'gradle';
+const GRADLE_URL_PREFIX = 'distributionUrl=';
+const GRADLE_URL_TEMPLATE = 'https\\://services.gradle.org/distributions/gradle-VERSION-all.zip';
+const GRADLE_MAX_ERROR_LOG_LINES = 15;
+
+const GRADLE_DEPENDENCIES_PLACEHOLDER = '// additionalAppDependencies placeholder (don\'t change or delete this line)';
+
+const VERSION_KEYS = [
+  GRADLE_VERSION_KEY,
+  'androidGradlePlugin',
+  'compileSdk',
+  'buildTools',
+  'minSdk',
+  'targetSdk',
+  'kotlin'
+];
+
+const gradleLog = logger.getLogger('Gradle');
+
+class ServerBuilder {
+  constructor (args = {}) {
+    this.serverPath = args.serverPath;
+    this.showGradleLog = args.showGradleLog;
+
+    const buildConfiguration = args.buildConfiguration || {};
+
+    const versionConfiguration = buildConfiguration.toolsVersions || {};
+    const serverVersions = {};
+    for (const key of VERSION_KEYS) {
+      if (versionConfiguration[key]) {
+        serverVersions[key] = versionConfiguration[key];
+      }
+    }
+    this.serverVersions = serverVersions;
+
+    this.additionalAppDependencies = buildConfiguration.additionalAppDependencies || [];
+  }
+
+  async build () {
+    if (this.serverVersions[GRADLE_VERSION_KEY]) {
+      await this.setGradleWrapperVersion(this.serverVersions[GRADLE_VERSION_KEY]);
+    }
+    await this.insertAdditionalAppDependencies(this.additionalAppDependencies);
+    await this.runBuildProcess();
+  }
+
+  getCommand () {
+    const cmd = system.isWindows() ? 'gradlew.bat' : './gradlew';
+    let args = VERSION_KEYS
+      .filter(key => key !== GRADLE_VERSION_KEY)
+      .map(key => {
+        const serverVersion = this.serverVersions[key];
+        const gradleProperty = `appium${key.charAt(0).toUpperCase()}${key.slice(1)}`;
+        return serverVersion ? `-P${gradleProperty}=${serverVersion}` : null;
+      })
+      .filter(Boolean);
+
+    args.push('assembleAndroidTest');
+
+    return {cmd, args};
+  }
+
+  async setGradleWrapperVersion (version) {
+    const propertiesPath = path.resolve(this.serverPath, 'gradle', 'wrapper', 'gradle-wrapper.properties');
+    const originalProperties = await fs.readFile(propertiesPath, 'utf8');
+    const newProperties = this.updateGradleDistUrl(originalProperties, version);
+    await fs.writeFile(propertiesPath, newProperties, 'utf8');
```

Comment:
should we check if newProperties is different from originalProperties first?

Reply:
This line will only be executed if some Gradle version was specified in a build configuration file. So I'm not sure such check is needed.